src/v/cluster/shard_placement_table.h

```diff
@@ -0,0 +1,172 @@
+/// independently.
+///
+/// Note that in contrast to `cluster::shard_table` (that helps other parts of
+/// the system to find the shard where the `cluster::partition` object for some
+/// ntp resides) this is a more granular table that is internal to the partition
+/// reconciliation process.
+class shard_placement_table
+  : public ss::peering_sharded_service<shard_placement_table> {
+public:
+    // assignment modification methods must be called on this shard
+    static constexpr ss::shard_id assignment_shard_id = 0;
+
+    enum class hosted_status {
+        /// Cross-shard transfer is in progress, we are the destination
+        receiving,
+        /// Normal state, we can start the partition instance.
+        hosted,
+        /// We have transferred our state to somebody else, now our copy must be
+        /// deleted.
+        obsolete,
+    };
+
+    /// Current state of shard-local partition kvstore data on this shard.
+    struct shard_local_state {
+        model::revision_id log_revision;
+        hosted_status status;
+
+        static shard_local_state initial(model::revision_id log_revision) {
+            return shard_local_state{
+              .log_revision = log_revision,
+              .status = hosted_status::hosted,
+            };
+        }
+
+        static shard_local_state receiving(model::revision_id log_revision) {
+            return shard_local_state{
+              .log_revision = log_revision,
+              .status = hosted_status::receiving,
+            };
+        }
+
+        friend std::ostream&
+        operator<<(std::ostream&, const shard_local_state&);
+    };
+
+    /// A struct holding both current shard-local and target states for an ntp.
+    struct placement_state {
+        std::optional<shard_local_state> local;
+        std::optional<shard_placement_target> target;
+        model::shard_revision_id shard_revision;
+
+        // invariant: if local is nullopt, then expected_on_this_shard() is true
+        // (there is no point in keeping an item that has no local state as well
+        // as is not expected on this shard)
+        bool expected_on_this_shard() const {
+            return !_next && target && target->shard == ss::this_shard_id();
+        }
+
+        friend std::ostream& operator<<(std::ostream&, const placement_state&);
+
+    private:
+        friend class shard_placement_table;
+
+        placement_state() = default;
+        placement_state(
+          const shard_placement_target& target, model::shard_revision_id rev)
+          : target(target)
+          , shard_revision(rev) {}
+
+        /// If this shard is the initial shard for this partition on this node,
+        /// this field will contain the corresponding shard revision.
+        model::shard_revision_id _is_initial_at_revision;
+        /// If x-shard transfer is in progress, will hold the destination.
+        std::optional<ss::shard_id> _next;
```

Comment:
isn't that equal to `target.shard` ?

Reply:
Its purpose is to hold the destination if the target changes (so that we can finish started transfer). I'll update the comment.